Adding gitpod, codespaces, devcontainer configs along with pre-commit code formatting using black and prettier (#264)
cytomining/pycytominer: CONTRIBUTING.md

[Quick links](#quick-links)

[How can I contribute?](#how-can-i-contribute)

> - [Bug reporting](#bug-reporting)
> - [Suggesting enhancements](#suggesting-enhancements)
> - [Dev Environments](#dev-environments)
> - [Your first code contribution](#your-first-code-contribution)
> - [Pull requests](#pull-requests)

Reviewer: ```suggestion
- [Bug reporting](#bug-reporting)
- [Suggesting enhancements](#suggesting-enhancements)
- [Your first code contribution](#your-first-code-contribution)
- [Pull requests](#pull-requests)
- [Dev Environments](#dev-environments)
```

I think this contribution is important, but I'd like to avoid overwhelming a newcomer to this project. It's possible to contribute to this codebase without dev environments (e.g. documentation PR), so listing this last would be my preference.

Reviewer: @kenibrewer - can you also make the necessary changes to restructure the order below?

Please provide your enhancement suggestions as an issue in the repository:

### Dev Environments

> Cloud development configurations have been set up with both [Github Codespaces](https://docs.github.com/en/codespaces/overview) and [GitPod](https://www.gitpod.io/docs/introduction). These development environments have been set up with the local copy of the repository installed in development mode along with the tools  `black`, `pytest`, `codecov`. Pre-installed git hooks will ensure that any changed code is properly formatted by `black` prior to commits.

Author: Given that there is a one-sentence-per-line structure that you wish to enforce for READMEs, I added the `prettier` pre-commit hook which handles markdown files. The updated pull request is an example of how pre-commit hooks work. They only ever scan files that are being changed, and ignore all other files.

Please provide your enhancement suggestions as an issue in the repository:

### Dev Environments

#### Manual Setup

> General guidance for setting up a local development environment in Linux are provided below.

Reviewer: ```suggestion
We provide general guidance for setting up a local development environment in Linux below.
```

Reviewer: active voice (compared to passive voice) to keep same style throughout document

> It is strongly recommended to perform this within a virtual environment such as [conda](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html) or [pyenv](https://akrabat.com/creating-virtual-environments-with-pyenv/):

Reviewer: ```suggestion
We strongly recommend performing this within a virtual environment such as [conda](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html) or [pyenv](https://akrabat.com/creating-virtual-environments-with-pyenv/):
```

```bash
# Checkout the repository
git clone https://github.com/cytomining/pycytominer.git
cd pycytominer
# Install pycytominer in development mode along with associated tools
bash .devcontainer/postCreateCommand.sh
```

#### Local DevContainer

Instructions for setting up a local development environment using VSCode DevContainers:

1. Install [VSCode](https://code.visualstudio.com/download)
2. Install the [Remote - Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extension
3. Open the repository in VSCode
4. Click on the green "Reopen in Container" button in the lower left corner of the window
5. Wait for the container to build and install the required dependencies

> #### Cloud Environment (Recommended)

Reviewer: ```suggestion
#### Cloud environment
```

We can add the `Recommended` flag to this header once we spend more time using this method :)

Feel free to add a more direct note re. recommended in the section below however. For example, might it be worth saying: "We recommend using the cloud development configuration over local dev containers"?

Also note the lower case header to keep consistent documentation

Author: Makes sense. Added a line recommending local/cloud dev containers over manual setup.

> Cloud development configurations have been set up with both [Github Codespaces](https://github.com/codespaces) and [GitPod](https://www.gitpod.io/).

Reviewer: ```suggestion
We've set up cloud development configurations with both [Github Codespaces](https://github.com/codespaces) and [GitPod](https://www.gitpod.io/).
```

Author: Done

> These development environments have been set up with the local copy of the repository installed in development mode along with the tools needed from `requirements-dev.txt`.
> Pre-installed git hooks will ensure that any changed code is properly formatted by required tools prior to commits.
> Using a pre-built cloud development environment is the easiest way to get started contributing to pycytominer and both Gitpod and Codespaces have generous free usage tiers.
> When you are ready to make a pull request, you can use the same cloud environment to run the full test suite and ensure that your changes pass all tests.
> These cloud environments can be launched by clicking on the following links:

Reviewer: ```suggestion
These development environments include the local copy of the repository installed in development mode along with the tools specified in `requirements-dev.txt`.
Prior to commit, pre-installed git hooks auto-format any changed code.
Using a pre-built cloud development environment is an easy way to get started contributing to pycytominer, and both Gitpod and Codespaces have generous free usage tiers.
When you are ready to make a pull request, please use the same cloud environment to run the full test suite and ensure that your changes pass all tests.
You can launch these cloud environments by clicking on the following links:
```

Minor formatting suggestions to maintain consistent tone. I don't think that I changed meaning, but please fix if you spot anything

Author: LGTM

Please provide your enhancement suggestions as an issue in the repository:

### Dev Environments

#### Manual Setup

Instructions for setting up a local development environment in a virtual environment:

```bash
# Checkout the repository
git clone https://github.com/cytomining/pycytominer.git
cd pycytominer

# Install pyenv if not already installed

> curl https://pyenv.run | bash

Author: I think it is worth providing some guidance on manual setup of development environments within this PR. If someone follows the recommended approach for their manual dev environment setup, it will put them on the same page in terms of auto-formatting of code contributions as people using devcontainers. I did however, add some caveats and clarifying language around this particular approach that makes it clear that the specific commands are intended for a linux dev environment and shifting the responsibility to virtual environment management back to the user.

These development environments have been set up with the local copy of the repository installed in development mode along with the tools needed from `requirements-dev.txt`.
Pre-installed git hooks will ensure that any changed code is properly formatted by required tools prior to commits.
Using a pre-built cloud development environment is the easiest way to get started contributing to pycytominer and both Gitpod and Codespaces have generous free usage tiers.
When you are ready to make a pull request, you can use the same cloud environment to run the full test suite and ensure that your changes pass all tests.
These cloud environments can be launched by clicking on the following links:

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://open.vscode.dev/cytomining/pycytominer)
[Beginner's Guide to Codespaces](https://github.blog/2023-02-22-a-beginners-guide-to-learning-to-code-with-github-codespaces/)
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/cytomining/pycytominer)

> [Beginner's Guide to Gitpod](https://www.gitpod.io/docs/introduction/getting-started)

Author: Done

For python code style, we use [black](https://github.com/psf/black).
Please use black before committing any code.
We will not accept code contributions that do not use black.

> If you have set up your development environment using one of the options above, required formatting tools will be installed and run automatically on any modified files before commits using [pre-commit](https://pre-commit.com/).

Reviewer: ```suggestion
If you have set up your development environment using one of the dev container options specified below, the containers will install all required formatting tools, which will run automatically on any modified files before commits (using a tool called [pre-commit](https://pre-commit.com/)).
```

Minor edits, one to incorporate the future change of adding the devcontainers below

### Python style guide

For python code style, we use [black](https://github.com/psf/black).

> For markdown style, we use [prettier](https://prettier.io/).

Reviewer 2: Would this be better situated in the `### Documentation style guide` section below?

Author: Done
